File: solutions/zig/kindergarten-garden/1/kindergarten_garden.zig
const std = @import("std");

pub const Plant = enum {
    clover,
    grass,
    radishes,
    violets,
};

fn charToPlant(c: u8) Plant {
    return switch (c) {
        'C' => .clover,
        'G' => .grass,
        'R' => .radishes,
        'V' => .violets,
        else => unreachable,
    };
}

pub fn plants(diagram: []const u8, student: []const u8) [4]Plant {
    const students = [_][]const u8{
        "Alice", "Bob",     "Charlie", "David",  "Eve",     "Fred",
        "Ginny", "Harriet", "Ileana",  "Joseph", "Kincaid", "Larry",
    };

    var student_index: usize = 0;
    for (students, 0..) |s, i| {
        if (std.mem.eql(u8, s, student)) {
            student_index = i;
            break;
        }
    }

    const offset = student_index * 2;
    const line_length = std.mem.indexOf(u8, diagram, "\n").?;

    const row1 = diagram[0..line_length];
    const row2 = diagram[line_length + 1 ..];

    return [_]Plant{
        charToPlant(row1[offset]),
        charToPlant(row1[offset + 1]),
        charToPlant(row2[offset]),
        charToPlant(row2[offset + 1]),
    };
}

File: solutions/typescript/kindergarten-garden/1/kindergarten-garden.ts
type Student = string;

const DEFAULT_STUDENTS: Student[] = [
    'Alice',
    'Bob',
    'Charlie',
    'David',
    'Eve',
    'Fred',
    'Ginny',
    'Harriet',
    'Ileana',
    'Joseph',
    'Kincaid',
    'Larry',
];

const PLANT_CODES = {
    G: 'grass',
    V: 'violets',
    R: 'radishes',
    C: 'clover',
} as const;

type Plant = (typeof PLANT_CODES)[keyof typeof PLANT_CODES];
type Plants = Plant[];

export class Garden {
    private diagram: string;
    private students: Student[];

    constructor(diagram: string, students = DEFAULT_STUDENTS) {
        this.diagram = diagram;
        this.students = students.sort();
    }

    public plants(student: Student): Plants {
        const offset = this.students.indexOf(student) * 2;

        const [row1, row2] = this.diagram.split('\n');

        return [
            PLANT_CODES[row1[offset] as keyof typeof PLANT_CODES],
            PLANT_CODES[row1[offset + 1] as keyof typeof PLANT_CODES],
            PLANT_CODES[row2[offset] as keyof typeof PLANT_CODES],
            PLANT_CODES[row2[offset + 1] as keyof typeof PLANT_CODES],
        ];
    }
}
